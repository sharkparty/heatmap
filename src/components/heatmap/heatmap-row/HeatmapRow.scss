$blue: rgba(5, 17, 129, .1);
$green: rgba(45, 190, 0, 1);
$yellow: rgba(228, 227, 0, 1);
$red: rgba(239, 83, 47, 1);

.heatmap-row {
  min-height: 20px;
  display: block;
  $blueMax: 34;
  transition: background-color 250ms;

  @for $i from 0 through $blueMax {
    &.blue-#{$i} {
      background-color: rgba(green, $i/$blueMax);
    }
  }

  $yellowMax: 67;
  @for $i from 33 through $yellowMax {
    &.yellow-#{$i} {
      background-color: rgba(yellow, $i/$yellowMax);
    }
  }

  $redMax: 102;
  @for $i from 66 through $redMax {
    &.red-#{$i} {
      background-color: rgba(red, $i/$redMax);
    }
  }

  &.empty {
    background-color: $blue;
  }
}
