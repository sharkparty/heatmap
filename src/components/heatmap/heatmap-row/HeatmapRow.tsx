// React
import * as React from 'react';
// Vendor
import * as moment from 'moment';
import * as classnames from 'classnames';
import { compose, withProps, ComponentEnhancer } from 'recompose';
// Assets
import './HeatmapRow.css';

export interface HeatmapRowParams {
    intervalOfTime: number;
    config: {};
    meta: {
        min?: number;
        max?: number;
        range: number;
    };
    label?: string;
    intervalData?: IntervalData;
    className?: string;
}

export interface IntervalData {
    value: number;
    label: string;
}

const defaultIntervalData: IntervalData = {
    value: 0,
    label: 'No data'
};
/**
 *
 * @param {number} range
 * @param {number} intervalOfTime
 * @param {IntervalData} intervalData
 * @returns [{ intervalOfTime: number; intervalData; label: string; className: string }]
 */
export const heatmapRowPropertyMapper = ({
    meta: { range },
    intervalOfTime,
    intervalData = defaultIntervalData
}: HeatmapRowParams) => {
    const { value }: IntervalData = intervalData;
    const label: string = `${moment().hour(intervalOfTime).format('ha')}: ${value || ''}`;
    const scaledValue: number = range ? Math.round((value / range) * 100) : value;
    let className: string = '';

    if (scaledValue === 0) {
        className = 'empty';
    } else if (scaledValue < 33.3) {
        className = `blue-${scaledValue}`;
    } else if (scaledValue < 66.6) {
        className = `yellow-${scaledValue}`;
    } else if (scaledValue > 66.6) {
        className = `red-${scaledValue}`;
    }

    return {
        intervalOfTime,
        intervalData,
        label,
        className
    };
};

export const enhance: ComponentEnhancer<{}, HeatmapRowParams> = compose(
    withProps(heatmapRowPropertyMapper)
);

/**
 * @param {number} intervalOfTime
 * @param {IntervalData} intervalData
 * @param {string} label
 * @param {string} className
 * @returns {any}
 * @constructor
 */
export const HeatmapRow: React.SFC<HeatmapRowParams> = ({
    intervalOfTime,
    intervalData,
    label,
    className
}) => {
    return (
        <div
            aria-label={label}
            title={label}
            className={classnames('heatmap-row', className)}
        />
    );
};

export default enhance(HeatmapRow);
