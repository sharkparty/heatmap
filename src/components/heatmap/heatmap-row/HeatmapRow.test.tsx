// React
import * as React from 'react';
// Vendor
import { shallow } from 'enzyme';
// Search
import { HeatmapRow, IntervalData, heatmapRowPropertyMapper } from './HeatmapRow';

describe('Component: HeatmapRow', () => {
    it('renders without crashing', () => {
        const wrapper = shallow(
            <HeatmapRow intervalOfTime={1} config={{}} meta={{ range: 1 }} />
        );
        expect(wrapper).toMatchSnapshot();
    });
    it('renders with meta', () => {
        const wrapper = shallow(
            <HeatmapRow intervalOfTime={1} config={{}} meta={{ min: 1, max: 2, range: 1 }} />
        );
        expect(wrapper).toMatchSnapshot();
    });
    describe('HOC: heatmapRowPropertyMapper', () => {
        let mockIntervalData: IntervalData;
        let mockMeta = { range: 100 };
        beforeEach(() => {
            mockIntervalData = {
                value: 10,
                label: 'Test'
            };
        });

        it('should return a blue class', () => {
            const subject = heatmapRowPropertyMapper({
                meta: mockMeta,
                intervalOfTime: 1,
                intervalData: mockIntervalData,
                config: {}
            });
            expect(subject.className).toBe('blue-10');
            expect(subject.intervalData.label).toBe('Test');
        });
        it('should return a yellow class', () => {
            mockIntervalData.value = 50;
            const subject = heatmapRowPropertyMapper({
                meta: mockMeta,
                intervalOfTime: 1,
                intervalData: mockIntervalData,
                config: {}
            });
            expect(subject.className).toBe('yellow-50');
            expect(subject.intervalData.label).toBe('Test');
        });
        it('should return a red class', () => {
            mockIntervalData.value = 99;
            const subject = heatmapRowPropertyMapper({
                meta: mockMeta,
                intervalOfTime: 1,
                intervalData: mockIntervalData,
                config: {}
            });
            expect(subject.className).toBe('red-99');
            expect(subject.intervalData.label).toBe('Test');
        });
    });
});
