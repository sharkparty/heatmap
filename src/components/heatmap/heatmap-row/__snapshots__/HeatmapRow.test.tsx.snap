// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: HeatmapRow renders with meta 1`] = `
<div
  className="heatmap-row"
/>
`;

exports[`Component: HeatmapRow renders without crashing 1`] = `
<div
  className="heatmap-row"
/>
`;
